### PR TITLE
78 vault dynamic db secrets

### LINKED ISSUE #78: Change Vault implementation to use more than KV secrets engine
As of now, ocelot only uses the key value secrets engine. We should be using the appropriate secrets engine for secret type. 

At the very least: 
[database engine](https://www.vaultproject.io/docs/secrets/databases/postgresql.html) for postgres retrieval 

---
this should be accompanied by an update of [go-til](https://github.com/shankj3/go-til)'s Vault implementation to also include token renewal. Using non-expiring tokens is just silly. When a new  instance of remoteConfig is instantiated, a thread should be kicked off to renew on an interval.

### PR DESCRIPTION
**Pull request by tjtelan**
Support addressing #78 

closes #78 

* Support for dynamic user/pass for postgres with Vault + auto renewing the lease when 75% of the time has elapsed.
* Added a lot more testing for remoteconfig.

Upgrade notes:
New consul values need to be set to choose between using statically set passwords stored in Vault (the current standard) or dynamically set via Vault's Database secret backend.

The testing around dynamic secrets coming soon, but for now, using the new Vagrantfile changes will demonstrate setting up consul + vault to use this new functionality.

Example output:
```
vagrant@ubuntu-bionic:~/go/src/github.com/shankj3/ocelot$ admin
{"function":"main.main","level":"debug","msg":":10000","time":"2018-10-19T18:29:13Z"}
postgres storage
{"function":"github.com/shankj3/ocelot/common/credentials.(*RemoteConfig).getForPostgres","level":"debug","msg":"Dynamic Postgres creds","time":"2018-10-19T18:29:13Z"}
{"function":"github.com/shankj3/ocelot/vendor/github.com/shankj3/go-til/vault.(*VaultyImpl).GetVaultSecret","level":"debug","msg":"Secret from Vault \u0026{bea56a70-c06d-4f91-737d-1054104e0084 database/creds/ocelot/545a3bd4-0286-60dc-aab4-7896f3e39f60 600 true map[password:A1a-6HzJOpPnrTNliTfP username:v-token-ocelot-4Xpj4P98q9BHbfJLkPC5-1539973753] [] \u003cnil\u003e \u003cnil\u003e}","time":"2018-10-19T18:29:13Z"}
{"function":"github.com/shankj3/ocelot/common/credentials.(*RemoteConfig).getForPostgres","level":"debug","msg":"Dynamic postgres creds from Vault: \u0026{bea56a70-c06d-4f91-737d-1054104e0084 database/creds/ocelot/545a3bd4-0286-60dc-aab4-7896f3e39f60 600 true map[password:A1a-6HzJOpPnrTNliTfP username:v-token-ocelot-4Xpj4P98q9BHbfJLkPC5-1539973753] [] \u003cnil\u003e \u003cnil\u003e}","time":"2018-10-19T18:29:13Z"}

{"function":"github.com/shankj3/ocelot/vendor/github.com/shankj3/go-til/vault.(*VaultyImpl).RenewLeaseForever","level":"debug","msg":"About to renew the lease for database/creds/ocelot/545a3bd4-0286-60dc-aab4-7896f3e39f60","time":"2018-10-19T18:36:43Z"}
{"function":"github.com/shankj3/ocelot/vendor/github.com/shankj3/go-til/vault.(*VaultyImpl).RenewLeaseForever","level":"debug","msg":"About to renew the lease for database/creds/ocelot/545a3bd4-0286-60dc-aab4-7896f3e39f60","time":"2018-10-19T18:44:13Z"}
{"function":"github.com/shankj3/ocelot/vendor/github.com/shankj3/go-til/vault.(*VaultyImpl).RenewLeaseForever","level":"debug","msg":"About to renew the lease for database/creds/ocelot/545a3bd4-0286-60dc-aab4-7896f3e39f60","time":"2018-10-19T18:51:43Z"}
{"function":"github.com/shankj3/ocelot/vendor/github.com/shankj3/go-til/vault.(*VaultyImpl).RenewLeaseForever","level":"debug","msg":"About to renew the lease for database/creds/ocelot/545a3bd4-0286-60dc-aab4-7896f3e39f60","time":"2018-10-19T18:59:13Z"}
{"function":"github.com/shankj3/ocelot/vendor/github.com/shankj3/go-til/vault.(*VaultyImpl).RenewLeaseForever","level":"debug","msg":"About to renew the lease for database/creds/ocelot/545a3bd4-0286-60dc-aab4-7896f3e39f60","time":"2018-10-19T19:06:43Z"}
{"function":"github.com/shankj3/ocelot/vendor/github.com/shankj3/go-til/vault.(*VaultyImpl).RenewLeaseForever","level":"debug","msg":"About to renew the lease for database/creds/ocelot/545a3bd4-0286-60dc-aab4-7896f3e39f60","time":"2018-10-19T19:14:13Z"}
```

### CODE REVIEW

File: common/credentials/remoteconfig.go [hunk 1]
```diff
@@ -101,15 +105,16 @@ func GetInstance(consulHost string, consulPort int, token string) (CVRemoteConfi
...
+	Location     string
+	Port         int
+	DbName       string
+	Password     string
+	VaultLeaseID string
 }
 
 type StorageCred interface {
-	GetStorageCreds(typ storage.Dest) (*StorageCreds, error)
+	GetStorageCreds(typ *storage.Dest) (*StorageCreds, error)
```
Comment (shankj3) [inline common/credentials/remoteconfig.go]: why this change? adding pointers and having to dereference for ints seems like overkill

File: common/credentials/remoteconfig.go [hunk 2]
```diff
@@ -420,72 +432,168 @@ func (rc *RemoteConfig) GetStorageType() (storage.Dest, error) {
...
+	}
+
+	//vaultRoleName := ""
+	vaultBackend := "kv" // The default backend is "kv" if left unconfigured in Consul
+	for _, vconf := range vaultConf {
+		switch vconf.Key {
+		case common.VaultDBSecretEngine:
+			switch string(vconf.Value) {
+			case "kv":
+				ocelog.Log().Debugf("Static Postgres creds")
```
Comment (shankj3) [inline common/credentials/remoteconfig.go]:
```suggestion
				ocelog.Log().Info("Static Postgres creds")
```

File: common/credentials/remoteconfig.go [hunk 3]
```diff
@@ -420,72 +432,168 @@ func (rc *RemoteConfig) GetStorageType() (storage.Dest, error) {
...
+				break
+			case "database":
+				ocelog.Log().Debugf("Dynamic Postgres creds")
```
Comment (shankj3) [inline common/credentials/remoteconfig.go]:
```suggestion
				ocelog.Log().Info("Dynamic Postgres creds")
```

File: common/credentials/remoteconfig.go [hunk 4]
```diff
@@ -420,72 +432,168 @@ func (rc *RemoteConfig) GetStorageType() (storage.Dest, error) {
...
+				break
+			default:
+				return &StorageCreds{}, errors.New("Unsupported Vault DB Secret Engine")
+			}
+			vaultBackend = string(vconf.Value)
+			//case common.VaultRoleName:
```
Comment (shankj3) [inline common/credentials/remoteconfig.go]: delete?

File: common/credentials/remoteconfig.go [hunk 5]
```diff
@@ -420,72 +432,168 @@ func (rc *RemoteConfig) GetStorageType() (storage.Dest, error) {
...
 	storeConfig := &StorageCreds{}
-	for _, pair := range pairs {
+
+	kvconfig, err := rc.Consul.GetKeyValues(common.PostgresCredLoc)
+	if len(kvconfig) == 0 || err != nil {
+		errorMsg := fmt.Sprintf("unable to get postgres creds from consul")
+		if err != nil {
+			errorMsg = fmt.Sprintf("%s, err: %s", errorMsg, err.Error())
+		}
+		return &StorageCreds{}, errors.New(errorMsg)
```
Comment (shankj3) [inline common/credentials/remoteconfig.go]: can just return nil because its a pointer returned
```suggestion
		return nil, errors.New(errorMsg)
```

File: common/credentials/remoteconfig.go [hunk 6]
```diff
@@ -420,72 +432,168 @@ func (rc *RemoteConfig) GetStorageType() (storage.Dest, error) {
...
-	if err != nil {
-		return storeConfig, errors.New("unable to get postgres password from vault, err: " + err.Error())
+
+	switch vaultBackend {
+	case "kv":
+		kvconfig, err := rc.Consul.GetKeyValues(common.PostgresCredLoc)
+		if len(kvconfig) == 0 || err != nil {
+			errorMsg := fmt.Sprintf("unable to get postgres location from consul")
+			if err != nil {
+				errorMsg = fmt.Sprintf("%s, err: %s", errorMsg, err.Error())
```
Comment (shankj3) [inline common/credentials/remoteconfig.go]: if you use `github.com/pkg/errors`, you can use `errors.Wrap(err, "msg")` and it is cleaner looking

File: common/credentials/remoteconfig.go [hunk 7]
```diff
@@ -420,72 +432,168 @@ func (rc *RemoteConfig) GetStorageType() (storage.Dest, error) {
...
 	if err != nil {
 		return nil, err
 	}
 	if typ == storage.Postgres {
 		fmt.Println("postgres storage")
 	}
-	creds, err := rc.GetStorageCreds(typ)
+
+	// This might return the full secrets struct
+	creds, err := rc.GetStorageCreds(&typ)
```
Comment (shankj3) [inline common/credentials/remoteconfig.go]: if you're trying to treat this integer type like a struct, might as well make methods off of it? it would get rid of this function and all the type switching, first step to cleaning up this admittedly stringy code
```golang
func (t typ) GetStorageCreds() {}
func (t typ) GetOcelotStorage() {}
```

File: common/credentials/remoteconfig.go (hunk 1 above)
Comment (tjtelan) [reply common/credentials/remoteconfig.go]: Unsurprisingly, this is a really pedantic change. This is a semantic from Rust that I am applying to Go. (The issue is [ownership](https://doc.rust-lang.org/book/second-edition/ch04-00-understanding-ownership.html)).
The code that used `GetStorageCreds(t)` (in `GetOcelotStorage()`) would not have compiled in Rust because we continue to read `t` after we pass ownership (non-pointer reference) of `t` to `GetOcelotStorage()`'s scope and into `GetStorageCreds(t)`, which is a potential data race condition.

More practically, I feel like this self-documents better as returning an internal attribute of a storage.Dest, instead of being a static interface method that takes a storage.Dest.

**Comment by tjtelan** [comment]
I've got this deployed in staging. Didn't catch the issues because the Vagrant env uses privileged users for postgresql and vault, but we properly limit permissions in staging/production.

Anyway, I'll take the other feedback and make changes. (Including offline feedback: Shelling out to call Vault instead of hitting api, so we can have a test that covers this new code) I just needed to get this deployed so you can see it in action.

**Comment by tjtelan** [comment]
I've got this deployed in staging, and its been running for more than 24 hrs w/o issues. Didn't catch the issues because the Vagrant env uses privileged users for postgresql and vault, but we properly limit permissions in staging/production.

Anyway, I'll be taking the other feedback now, and make changes. (Including offline feedback: Shelling out to call Vault instead of hitting api, so we can have a test that covers this new code) I just needed to get this deployed so you can see it in action.

**Comment by tjtelan** [comment]
It seems that I can't use the Vault cli to hit the in-mem instance from TestSetupVaultAndConsul(*testing.T), and this is really quite annoying. I even tried to set the client timeout to something absurdly long... Back to square one with setting up the database backend for tests.

ex.
```
$ VAULT_CLIENT_TIMEOUT=360 vault status
Error checking seal status: Get http://127.0.0.1:60092/v1/sys/seal-status: net/http: request canceled (Client.Timeout exceeded while awaiting headers)```

**Comment by tjtelan** [comment]
This is already turning out to be a really expensive test. And to make it even more tedious, I'm going to write this against a the vagrant `infra` box, just so we can have something predictable that tests the dynamic path.

File: storage/testutil.go [hunk 8]
```diff
@@ -78,7 +79,7 @@ func CreateTestPgDatabase(t *testing.T, port int) (cleanup func(t *testing.T), p
 	}
 	//var containerId string
 	//containerId = strings.Trim(outbe.String(), "\n")
-	t.Log("successfully started up test pg database on port 5555")
+	t.Log("successfully started up test pg database on port " + fmt.Sprintf("%d", port))
```
Comment (shankj3) [inline storage/testutil.go]:
```suggestion
	t.Logf("successfully started up test pg database on port %d", port)
```
